cmd/deploy.go

```diff
@@ -15,107 +15,395 @@
 		if err != nil {
-			//try and get the exposed ports and use the first one
-			Warning.log("Could not detect a container port (PORT env var).")
-			portsStr, portsErr := getExposedPorts()
-			if portsErr != nil {
-				return portsErr
+			Warning.log("Failed to find Appsody operator. Attempting to install...")
+			err = installCmd.RunE(cmd, args)
+			if err != nil {
+				return errors.Errorf("Failed to install Appsody operator. Exiting... %s", configFile)
 			}
-			if len(portsStr) == 0 {
-				//No ports exposed
-				Warning.log("This container exposes no ports. The service will not be accessible.")
-				port = 0 //setting this to 0
-			} else {
-				portStr := portsStr[0]
-				Warning.log("Picking the first exposed port as the KNative service port. This may not be the correct port.")
-				port, err = strconv.Atoi(portStr)
-				if err != nil {
-					Warning.log("The exposed port is not a valid integer. The service will not be accessible.")
-					port = 0
+		}
+
+		exists, err := exists(configFile)
+		if err != nil {
+			return errors.Errorf("Error checking status of %s", configFile)
+		}
+		if !exists || (exists && force == true) {
+			err = generateDeploymentConfig()
+			if err != nil {
+				if err.Error() == "docker cp command failed: exit status 1" {
+					Warning.log("No deployment config is present in the stack. Falling back to default deploy config using Knative.")
+					return deployWithKnative(cmd, args)
 				}
+				return err
 			}
 		}
-		//Get the KNative template file
-		knativeTempl := getKNativeTemplate()
+		Info.log("Found existing deployment manifest ", configFile)
+
+		// Extract code and build the image - and tags it if -t is specified
+		buildErr := buildCmd.RunE(cmd, args)
+		if buildErr != nil {
+			return buildErr
+		}
 
 		//Retrieve the project name and lowercase it
 		projectName, perr := getProjectName()
 		if perr != nil {
 			return errors.Errorf("%v", perr)
 		}
-		//Get the project name and make it the KNative service name
-		serviceName := projectName
 		deployImage := projectName // if not tagged, this is the deploy image name
 		if tag != "" {
 			deployImage = tag //Otherwise, it's the tag
 		}
-		// We're not pushing to a repository, so we need to use dev.local for Knative to be able to find it
-		if !push {
-			localtag := "dev.local/" + projectName
-			// Tagging the image using the tag as the deployImage for KNative
-			err = DockerTag(deployImage, localtag)
-			if err != nil {
-				return errors.Errorf("Tagging the image failed - exiting. Error: %v", err)
+		// Edit the deployment manifest to reflect the new tag
+		yamlFile, err := os.Open(configFile)
+		if err != nil {
+			if os.IsNotExist(err) {
+				return errors.Errorf("Config file does not exist %s. ", configFile)
+			}
+			return errors.Errorf("Failed reading file %s", configFile)
+		}
+		defer yamlFile.Close()
+		scanner := bufio.NewScanner(yamlFile)
+		scanner.Split(bufio.ScanLines)
+		var txtlines []string
+		for scanner.Scan() {
```

Comment:
I assume the yaml structure is too complicated to create structs to parse it like we do the index.yaml files?

Reply:
The challenge is that the CRD spec for the app-deploy.yaml is currently v1alpha1 so it likely to change (whereas the use of the specific `applicationImage:` field is unlikely to).

Ultimately we'd like to import the YAML definition as a struct from the appsody-operator project and then parse it using that directly.